python/ql/src/semmle/python/regex.qll

```diff
@@ -121,8 +121,57 @@ deprecated string mode_from_mode_object(Value obj) {
 abstract class RegexString extends Expr {
   RegexString() { (this instanceof Bytes or this instanceof Unicode) }
 
+  override string toString() {
+    result = this.(Bytes).getText()
+    or
+    result = this.(Unicode).getText()
+  }
+
+  /** result is true for those start chars that actually mark a start of a char set. */
+  boolean char_set_start(int pos) {
+    exists(int index |
+      char_set_delimiter(index, pos) = true and
+      (
+        index = 1 and result = true // if a '[' is first in the string (among brackets), it starts a char set
+        or
+        index > 1 and
+        not char_set_delimiter(index - 1, _) = false and
```

Comment:
I'm a bit curious about this line. Why not just `char_set_delimiter(index - 1, _) = true`? As far as I can tell, `char_set_delimiter` cannot fail for `index - 1` if it has already succeeded for `index` (though I may have missed something).

Also, it would be great if each disjunct had a brief comment about which case it is handling.